### DISCUSSION TITLE
how to position rectangle around text?

### DISCUSSION THREAD

**Discussion by c-robinson** [tinne26/etxt#6] [Q&A]
I'm sorry I am having a bear of a time figuring out how to properly position a highlight around a word at an arbitrary point in the text. The rect returned by `SelectionRect(someText).ImageRect()` is the correct size but anchored at 0,0 not the position of the text. I then tried 

```
	activeHighlight = r.engine.SelectionRect(element).ImageRect()
	pt := image.Pt(feed.Position.X.Round(), feed.Position.Y.Round())
	activeHighlight = activeHighlight.Add(pt)
```

but that is positioning the highlight _under_ the target text.

So I think I have two questions:
1. what is the correct way to ensure my rectangle is properly anchored around the text element, or at least to find the upper-left corner of `someText`?
2. would it be possible to have the x and y `SelectionRect(someText).ImageRect()` be anchored to the word instead of at 0,0?

Also, you might want to put some contact information in the README, or add your email to your account on github

**Comment by tinne26** [comment]
I'm not sure I understand what you want to do. I don't know if you are trying to create a "selection highlight rectangle" behind the text, or some kind of box or arrow above or below it. Assuming the first:
- The feed's X and Y position is the pen position, which points to the *baseline* of the text. And when you are done drawing or advancing a word, you will be at the baseline position *at the end* of that word. So if you measure that word with `SelectionRect()`, you would need to subtract the width and the ascent.

If you are not familiar with some of these metrics, I highly recommend the document linked in the README. This is the most relevant part for your purposes: <https://freetype.org/freetype2/docs/glyphs/glyphs-3.html>.

Now, the ascent metric is not super easy to access through etxt. The renderer actually stores the value cached, so I may want to add a method for it to make it more efficient and straightforward, but for the moment you have to do something like this:
```Golang
font := renderer.GetFont()
size := renderer.GetSizePxFract()
metrics := renderer.GetSizer().Metrics(font, size)
ascent := metrics.Ascent
```

Regarding my contact, I don't want to share. Opening an issue even for a question like this is ok to me, but I've also opened discussions in case people want to ask small questions like these and opening an issue feels wrong for them.

**Comment by tinne26** [comment]
There are many other ways to deal with this though. For example, if you start the `Feed` with `Top` vertical align, you could manually increase the y position by the line advance each time you find a line break, and then you would have the "top" position of the current line, which you could use instead of the ascent.

About returning the text origin position from `SelectionRect()`, I rejected that because then:
- You have to pass a position to `SelectionRect()`, which in many cases is unnecessary.
- If you pass the position, do you need two versions, one with `int` and one with `fixed.Int26_6`?
- The align of the renderer would suddenly matter for `SelectionRect()`.
- The resulting rect would be a more complex struct.
- You can implement it with a wrapper or helper function if you really want it anyway.

**Comment by c-robinson** [comment]
> Regarding my contact, I don't want to share. Opening an issue even for a question like this is ok to me, but I've also opened discussions in case people want to ask small questions like these and opening an issue feels wrong for them.

No problem, totally understand! It's just in your README you advise people to reach out to you and I couldn't figure out how to do that without going to an issue, which as you noted, feels weird for this kind of question. Really appreciate you firing up discussions for this codebase!
